diff/diff.1
.\"
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	from 4.3BSD diff.1	6.4 (Berkeley) 5/19/86
.\"
.\" This code contains changes by
.\" Gunnar Ritter, Freiburg i. Br., Germany, March 2003. All rights reserved.
.\"
.\" Conditions 1, 2, and 4 and the no-warranty notice below apply
.\" to these changes.
.\"
.\"
.\" Copyright (c) 1980, 1993
.\"      The Regents of the University of California.  All rights reserved.
.\"
.\" SPDX-Licence-Identifier: BSD-4-Clause-UC
.\"
.\"
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH DIFF 1 "6/28/05" "Heirloom Toolchest" "User Commands"
.SH NAME
diff \- differential file comparator
.SH SYNOPSIS
.HP
.nh
.ad l
\fBdiff\fR
[\fB\-abBiptw\fR]
[\fB\-cefhnu\fR]
[\fB\-C\ \fInumber\fR]
[\fB\-U\ \fInumber\fR]
\fIfile1\fR \fIfile2\fR
.HP
.nh
.ad l
\fBdiff\fR
[\fB\-abBiptw\fR]
[\fB\-D\ \fIstring\fR]
\fIfile1\fR \fIfile2\fR
.HP
.nh
.ad l
\fBdiff\fR
[\fB\-abBiNptw12\fR]
[\fB\-cefhnu\fR]
[\fB\-C\ \fInumber\fR]
[\fB\-U\ \fInumber\fR]
[\fB\-lrs\fR]
[\fB\-S\ \fIname\fR]
[\fB\-x\ \fIpattern\fR]
[\fB\-X\ \fIname\fR]
\fIdirectory1\fR \fIdirectory2\fR
.br
.hy 1
.SH DESCRIPTION
.I Diff
tells what lines must be changed in two files to bring them
into agreement.
If
.I file1
.RI ( file2 )
is `\-', the standard input is used.
If
.I file1
.RI ( file2 )
is a directory, then a file in that directory
whose file-name is the same as the file-name of
.I file2
.RI ( file1 )
is used.
The normal output contains lines of these forms:
.IP "" 5
.I n1
a
.I n3,n4
.br
.I n1,n2
d
.I n3
.br
.I n1,n2
c
.I n3,n4
.PP
These lines resemble
.I ed
commands to convert
.I file1
into
.IR file2 .
The numbers after the letters pertain to
.IR file2 .
In fact, by exchanging `a' for `d' and reading backward
one may ascertain equally how to convert
.I file2
into
.IR file1 .
As in
.I ed,
identical pairs where
.I n1
=
.I n2
or
.I n3
=
.I n4
are abbreviated as a single number.
.PP
Following each of these lines come all the lines that are
affected in the first file flagged by `<',
then all the lines that are affected in the second file
flagged by `>'.
.TP 10
.B \-a
causes a list of differences to be output
for all files,
even for those found to have binary content.
This option is an extension.
.TP 10
.B \-b
causes trailing whitespace characters
to be ignored, and other
strings of whitespace to compare equal.
.TP 10
.B \-i
ignores the case of letters.  E.g., ``A'' will compare equal to ``a''.
.TP 10
.B \-p
causes the name of the surrounding C function,
or, more exactly,
of the first previous unchanged line
beginning with a letter, the dollar sign, or the underscore,
to be output with each set of changes.
Implies
.I \-c
unless
.I \-u
is also present.
This option is an extension.
.TP 10
.B \-t
will expand tabs in output lines.  Normal,
.B \-c
or
.B \-u
output adds character(s) to the front of each line which may screw up
the indentation of the original source lines and make the output listing
difficult to interpret.  This option will preserve the original source's
indentation.
.TP 10
.B \-w
is similar to
.B \-b
but causes whitespace characters
to be totally ignored.
E.g., ``if\ (\ a\ ==\ b\ )'' will compare equal to ``if(a==b)''.
.TP 10
.B \-B
causes changes that consist entirely of empty lines added or deleted
to be ignored.
This option is an extension.
.PP
The following options are mutually exclusive:
.TP 10
.B \-c
produces a diff with three lines of context.
With
.B \-c
the output format is modified slightly:
the output beginning with identification of the files involved and
their creation dates and then each change is separated
by a line with a dozen *'s.
The lines removed from
.I file1
are marked with `\(mi '; those added to
.I file2
are marked `+ '.  Lines which are changed from one
file to the other are marked in both files with with `! '.
.\".sp
.\"Changes which lie within <context> lines of each other are grouped
.\"together on output.  (This is a change from the previous ``diff -c''
.\"but the resulting output is usually much easier to interpret.)
.TP 10
\fB\-C\ \fInumber\fR
Same as
.B \-c
but uses
.I number
of lines of context.
.TP 10
\fB\-D\ \fIstring\fR
causes
.I diff
to create a merged version of
.I file1
and
.I file2
on the standard output, with C preprocessor controls included so that
a compilation of the result without defining \fIstring\fR is equivalent
to compiling
.I file1,
while defining
.I string
will yield
.I file2.
.TP 10
.B \-e
produces a script of
.I "a, c"
and
.I d
commands for the editor
.I ed,
which will recreate
.I file2
from
.IR file1 .
In connection with
.BR \-e ,
the following shell program may help maintain
multiple versions of a file.
Only an ancestral file ($1) and a chain of
version-to-version
.I ed
scripts ($2,$3,...) made by
.I diff
need be on hand.
A `latest version' appears on
the standard output.
.IP
\ \ \ \ \ \ \ \ (shift; cat $*; echo \'1,$p\') \(bv ed \- $1
.IP
Extra commands are added to the output when comparing directories with
.B \-e,
so that the result is a
.IR sh (1)
script for converting text files which are common to the two directories
from their state in
.I dir1
to their state in
.I dir2.
.TP 10
.B \-f
produces a script similar to that of
.B \-e,
not useful with
.I ed,
and in the opposite order.
.TP 10
.B \-h
does a fast, half-hearted job.
It works only when changed stretches are short
and well separated,
but does work on files of unlimited length.
.TP 10
.B \-n
produces a script similar to that of
.B \-e,
but in the opposite order and with a count of changed lines on each
insert or delete command.
.\"This is the form used by
.\".IR rcsdiff (1).
.TP 10
.B \-u
produces a unified diff with three lines of context.
The output begins with identification of the files involved
and their creation dates,
followed by the changes
separated by `@@ \-range +range @@'.
Lines removed from
.I file1
are marked with `\(mi',
those added to
.I file2
are marked `+'.
This option is an extension.
.TP 10
\fB\-U\ \fInumber\fR
Same as
.B \-u
but uses
.I number
of lines of context.
This option is an extension.
.PP
If both arguments are directories,
.I diff
sorts the contents of the directories by name, and then runs the
regular file
.I diff
algorithm on text files which are different.
Binary files which differ,
common subdirectories, and files which appear in only one directory
are listed.
.PP
Options when comparing directories are:
.TP 10
.B \-l
long output format; each text file
.I diff
is piped through
.IR pr (1)
to paginate it,
other differences are remembered and summarized
after all text file differences are reported.
.TP 10
.B \-N
causes the text of files
that exist in one directory only
to be output
as if compared to an empty file modified at 1/1/70.
This option is an extension.
.TP 10
.B \-1
is similar to
.IR \-N ,
but causes just the text of files that exist in
.I directory1
only to be output.
Files that exist only in
.I directory2
are listed.
This option is an extension.
.TP 10
.B \-2
is similar to
.IR \-N ,
but causes just the text of files that exist in
.I directory2
only to be output.
Files that exist only in
.I directory1
are listed.
This option is an extension.
.TP 10
.B \-r
causes application of
.I diff
recursively to common subdirectories encountered.
.TP 10
.B \-s
causes
.I diff
to report files which are the same, which are otherwise not mentioned.
.TP 10
.BI \-S " name"
starts a directory
.I diff
in the middle beginning with file
.I name.
.TP 10
.BI \-x " pattern"
excludes all file names that match
.I pattern
(as described in
.IR glob (7))
from comparison.
If
.I pattern
matches a directory,
files below that directory are also excluded.
This option is an extension.
.TP 10
.BI \-X " name"
excludes all file names contained in
.IR name .
This option is an extension.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale(7).
.TP
.B LC_CTYPE
Sets the mapping of bytes to characters,
character case translation
and the set of whitespace characters.
.TP
.B SYSV3
If this variable is set,
invalid options are ignored instead of being rejected,
and the text of some diagnostic messages is changed.
.SH FILES
.ta \w'/usr/5lib/diffh   'u
/tmp/d?????
.br
/usr/5lib/diffh	for \fB\-h\fR
.br
diff	for directory diffs
.br
pr
.SH "SEE ALSO"
bdiff(1),
cmp(1),
cc(1),
comm(1),
ed(1),
diff3(1),
patch(1),
locale(7)
.SH DIAGNOSTICS
Exit status is 0 for no differences, 1 for some, 2 for trouble.
.SH NOTES
Editing scripts produced under the
.BR \-e " or"
.BR \-f " option are naive about"
creating lines consisting of a single `\fB.\fR'.
.PP
When comparing directories with the
\fB\-b, \-w\fP, or \fB\-i\fP
options specified,
.I diff
first compares the files ala
.I cmp,
and then decides to run the
.I diff
algorithm if they are not equal.
This may cause a small amount of spurious output if the files
then turn out to be identical because the only differences are
insignificant blank string or case differences.
.PP
When
.I diff
output is used with
.IR ed (1)
or
.IR patch (1)
for file synchronization,
it is recommended that it is run in the
.I C
or another single-byte LC_CTYPE locale
since character-to-byte conversion
might otherwise hide some changes.


CHANGES
Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.

diff/diff.c
/*
 * This code contains changes by
 * Gunnar Ritter, Freiburg i. Br., Germany, March 2003. All rights reserved.
 *
 * Conditions 1, 2, and 4 and the no-warranty notice below apply
 * to these changes.
 *
 *
 * Copyright (c) 1991
 *      The Regents of the University of California.  All rights reserved.
 *
 * SPDX-Licence-Identifier: BSD-4-Clause-UC
 *
 *
 * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
 *
 * SPDX-Licence-Identifier: Caldera
 */

/*	Sccsid @(#)diff.c	1.24 (gritter) 3/27/05>	*/
/*	from 4.3BSD diff.c 4.6 4/3/86	*/

#include "diff.h"
#include <unistd.h>
#include <locale.h>
#include <iblok.h>

/* globals moved here from diff.h to prevent duplicated definitions */
int opt;
int	aflag;			/* diff binary files */
int	tflag;			/* expand tabs on output */
int	pflag;			/* show surrounding C function */
int	hflag;			/* -h, use halfhearted DIFFH */
int	bflag;			/* ignore blanks in comparisons */
int	wflag;			/* totally ignore blanks in comparisons */
int	iflag;			/* ignore case in comparisons */
int	Bflag;			/* ignore changes that consist of blank lines */
int	lflag;			/* long output format with header */
int	rflag;			/* recursively trace directories */
int	sflag;			/* announce files which are same */
int	Nflag;			/* write text of nonexistant files */
const char	*start;		/* do file only if name >= this */
struct xclusion *xflag;
int	wantelses;		/* -E */
char	*ifdef1;		/* String for -1 */
char	*ifdef2;		/* String for -2 */
char	*endifname;		/* What we will print on next #endif */
int	inifdef;
int context;
int status;
int	anychange;
char	*tempfile1;		/* used when comparing against std input */
char	*tempfile2;		/* used when comparing against std input */
char	**diffargv;		/* option list to pass to recursive diffs */
int	recdepth;		/* recursion depth */
jmp_buf	recenv;			/* jump stack on error */

struct stackblk *curstack;
char	*file1, *file2, *efile1, *efile2;
struct	stat stb1, stb2;
int mb_cur_max;
int sysv3;


/*
 * diff - driver and subroutines
 */

const char	diff[] = "diff";
const char	diffh[] = DIFFH;
const char	pr[] = "pr";
const char	*progname;
const char	*argv0;

static void	usage(void);
static void	xadd(const char *);
static void	Xadd(const char *);

int
main(int argc, char **argv)
{
	int	i, invalid = 0;

	progname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	if (getenv("SYSV3") != NULL)
		sysv3 = 1;
	ifdef1 = "FILE1"; ifdef2 = "FILE2";
	status = 2;
	argv0 = argv[0];
	diffargv = argv;
	while ((i = getopt(argc, argv, ":D:efnbBwitcC:uU:hS:rslNx:a12pX:"))
			!= EOF) {
		switch (i) {
#ifdef notdef
		case 'I':
			opt = D_IFDEF;
			wantelses = 0;
			break;
		case 'E':
			opt = D_IFDEF;
			wantelses = 1;
			break;
		case '1':
			opt = D_IFDEF;
			ifdef1 = optarg;
			break;
#endif
		case 'D':
			/* -Dfoo = -E -1 -2foo */
			wantelses = 1;
			ifdef1 = "";
			/* fall through */
#ifdef notdef
		case '2':
#endif
			opt = D_IFDEF;
			ifdef2 = optarg;
			break;
		case 'e':
			opt = D_EDIT;
			break;
		case 'f':
			opt = D_REVERSE;
			break;
		case 'n':
			opt = D_NREVERSE;
			break;
		case 'b':
			bflag = 1;
			break;
		case 'B':
			Bflag = 1;
			break;
		case 'w':
			wflag = 1;
			break;
		case 'i':
			iflag = 1;
			break;
		case 't':
			tflag = 1;
			break;
		case 'c':
			opt = D_CONTEXT;
			context = 3;
			break;
		case 'C': {
			char	*x;

			opt = D_CONTEXT;
			context = strtol(optarg, &x, 10);
			if (*x != '\0' || *optarg == '+' || *optarg == '-') {
				fprintf(stderr, "%s: use -C num\n", progname);
				done();
			}
			break;
		}
		case 'u':
			opt = D_UNIFIED;
			context = 3;
			break;
		case 'U': {
			char	*x;

			opt = D_UNIFIED;
			context = strtol(optarg, &x, 10);
			if (*x != '\0' || *optarg == '+' || *optarg == '-') {
				fprintf(stderr, "%s: use -U num\n", progname);
				done();
			}
			break;
		}
		case 'h':
			hflag++;
			break;
		case 'S':
			start = optarg;
			break;
		case 'r':
			rflag++;
			break;
		case 's':
			sflag++;
			break;
		case 'l':
			lflag++;
			break;
		case 'N':
			Nflag |= 3;
			break;
		case '1':
			Nflag |= 1;
			break;
		case '2':
			Nflag |= 2;
			break;
		case 'x':
			xadd(optarg);
			break;
		case 'a':
			aflag++;
			break;
		case 'p':
			pflag++;
			break;
		case 'X':
			Xadd(optarg);
			break;
		default:
			if (invalid == 0 && !sysv3)
				invalid = optopt;
		}
	}
	argv += optind, argc -= optind;
	if (argc != 2) {
		fprintf(stderr, sysv3 ? "%s: arg count\n" :
				"%s: two filename arguments required\n",
				progname);
		done();
	}
	file1 = argv[0];
	file2 = argv[1];
	if (invalid) {
		fprintf(stderr, "%s: invalid option -%c\n", progname, invalid);
		usage();
	}
	if (pflag) {
		if (opt == D_UNIFIED || opt == D_CONTEXT)
			/*EMPTY*/;
		else if (opt == 0) {
			opt = D_CONTEXT;
			context = 3;
		} else {
			fprintf(stderr,
				"%s: -p doesn't support -e, -f, -n, or -I\n",
		    		progname);
			done();
		}
	}
	if (hflag && opt) {
		fprintf(stderr,
		    "%s: -h doesn't support -e, -f, -n, -c, -u, or -I\n",
		    	progname);
		done();
	}
	diffany(argv);
	/*NOTREACHED*/
	return 0;
}

void
diffany(char **argv)
{
	if (!strcmp(file1, "-"))
		stb1.st_mode = S_IFREG;
	else if (stat(file1, &stb1) < 0) {
		if (sysv3)
			stb1.st_mode = S_IFREG;
		else {
			fprintf(stderr, "%s: %s: %s\n", progname, file1,
					strerror(errno));
			done();
		}
	}
	if (!strcmp(file2, "-"))
		stb2.st_mode = S_IFREG;
	else if (stat(file2, &stb2) < 0) {
		if (sysv3)
			stb2.st_mode = S_IFREG;
		else {
			fprintf(stderr, "%s: %s: %s\n", progname, file2,
					strerror(errno));
			done();
		}
	}
	if ((stb1.st_mode & S_IFMT) == S_IFDIR &&
	    (stb2.st_mode & S_IFMT) == S_IFDIR) {
		diffdir(argv);
	} else
		diffreg();
	done();
}

static void
usage(void)
{
	fprintf(stderr, "\
usage: %s [ -bcefhilnrstw -C num -D string -S name ] file1 file2\n",
		progname);
	done();
}

int
min(int a,int b)
{

	return (a < b ? a : b);
}

int
max(int a,int b)
{

	return (a > b ? a : b);
}

void
done(void)
{
	if (tempfile1) {
		unlink(tempfile1);
		tempfile1 = NULL;
	}
	if (tempfile2) {
		unlink(tempfile2);
		tempfile2 = NULL;
	}
	if (recdepth == 0)
		exit(status);
	else
		longjmp(recenv, 1);
}

static void	noroom(void);

void *
dalloc(size_t n)
{
	struct stackblk *sp;

	if ((sp = malloc(n + sizeof *sp)) != NULL) {
		sp->s_prev = NULL;
		sp->s_next = curstack;
		if (curstack)
			curstack->s_prev = sp;
		curstack = sp;
		return (char *)sp + sizeof *sp;
	} else
		return NULL;
}

void *
talloc(size_t n)
{
	void *p;

	if ((p = dalloc(n)) == NULL)
		noroom();
	return p;
}

void *
ralloc(void *p,size_t n)
{
	struct stackblk	*sp, *sq;

	if (p == NULL)
		return talloc(n);
	sp = (struct stackblk *)((char *)p - sizeof *sp);
	if ((sq = realloc(sp, n + sizeof *sp)) == NULL)
		noroom();
	if (sq->s_prev)
		sq->s_prev->s_next = sq;
	if (sq->s_next)
		sq->s_next->s_prev = sq;
	if (curstack == sp)
		curstack = sq;
	return (char *)sq + sizeof *sq;
}

void
tfree(void *p)
{
	struct stackblk	*sp;

	if (p == NULL)
		return;
	sp = (struct stackblk *)((char *)p - sizeof *sp);
	if (sp->s_prev)
		sp->s_prev->s_next = sp->s_next;
	if (sp->s_next)
		sp->s_next->s_prev = sp->s_prev;
	if (sp == curstack)
		curstack = sp->s_next;
	free(sp);
}

void
purgestack(void)
{
	struct stackblk *sp = curstack, *sq = NULL;

	do {
		free(sq);
		sq = sp;
		if (sp)
			sp = sp->s_next;
	} while (sq);
}

static void
noroom(void)
{
	oomsg(": files too big, try -h\n");
	status = 2;
	done();
}

static void
xadd(const char *cp)
{
	struct xclusion	*xp;

	xp = talloc(sizeof *xp);
	xp->x_pat = cp;
	xp->x_nxt = xflag;
	xflag = xp;
}

static void
Xadd(const char *name)
{
	struct iblok	*ip;
	char	*line = NULL;
	size_t	size = 0, len;

	if (name[0] == '-' && name[1] == '\0')
		ip = ib_alloc(0, 0);
	else
		ip = ib_open(name, 0);
	if (ip == NULL) {
		fprintf(stderr, "%s: -X %s: %s\n", progname, name,
				strerror(errno));
		done();
	}
	while ((len = ib_getlin(ip, &line, &size, realloc)) != 0) {
		if (line[len-1] == '\n')
			line[--len] = '\0';
		xadd(line);
		line = NULL;
		size = 0;
	}
	free(line);
	if (ip->ib_fd)
		ib_close(ip);
	else
		ib_free(ip);
}

void
oomsg(const char *s)
{
	write(2, progname, strlen(progname));
	write(2, s, strlen(s));
}
